upsilonconf/config.py

```diff
@@ -447,73 +449,83 @@ def to_dict(self, key_mods: Mapping[str, str] = None) -> Dict[str, Any]:
-    r : str
-        The replacement string.
+        The mapping object (`Configuration`, `dict`, ...) whose keys are to be
+        modified.
+    key_mods : Mapping
+        The dictionary with the replacements: All key strings are replaced
+        with the corresponding values from the this dictionary.
 
     Returns
     -------
-        The dictionary with all strings `s` replaced with `r` in the keys.
+    dict
+        A dictionary with the modified keys.
+
     """
-    # Traverse all the keys and replace `s` with `r`
-    dictionary = {}
-    for key, value in mapping.items():
-        try:
-            # Call this method recursively
-            value = __replace_in_keys(value, s, r)
-        except AttributeError:
-            # `value` is not of the mapping type
-            pass
-        dictionary[key.replace(s, r)] = value
+    # Build and compile replacement pattern
+    pattern = re.compile(
+        "|".join(
+            [re.escape(k) for k in sorted(key_mods, key=lambda k: len(k), reverse=True)]
+        )
+    )
 
-    return dictionary
+    return __modify_keys(mapping, key_mods, pattern)
 
 
-def _replace_in_keys(
-    config: Mapping[str, Any], key_modifiers: Mapping[str, str]
-) -> Mapping[str, Any]:
+def __modify_keys(mapping: Mapping[str, Any], key_mods: Mapping[str, str], pattern):
     """
     Replace strings in the keys of a mapping object.
 
+    The replacement is done recursively.
+
     Parameters
     ----------
-    config : Mapping
-        The configuration object (or dictionary) to be modified.
-    key_modifiers : dict
-        The dictionary with the replacements.
+    mapping : Mapping
+        The mapping object (`Configuration`, `dict`, ...) whose keys are to be
+        modified.
+    key_mods : Mapping
+        The dictionary with the replacements: All key strings are replaced
+        with the corresponding values from the this dictionary.
+    pattern : ???
+        The compiled replacement pattern.
 
     Returns
     -------
     dict
         A dictionary with the modified keys.
 
     """
-    _config = config
-    # Replace longest strings first
-    # - `sorted(..., reverse=True)` takes care of that
-    for key in key_modifiers.keys():
-        _config = __replace_in_keys(_config, key, key_modifiers[key])
-    return dict(_config)
+    dictionary = {}
+
+    # `mapping.items()` fails with `AttributeError`, if `mapping` is not of a
+    # mapping type, thus breaking the recursive calls
+    for key, value in mapping.items():
+        try:
+            # Call this method recursively
+            value = __modify_keys(value, key_mods, pattern)
+        except AttributeError:
+            # `value` is not of the mapping type
+            pass
+
+        # Replace only, if there are replacements requested, otherwise just
+        # save the value
+        if key_mods:
+            dictionary[pattern.sub(lambda m: key_mods[m.group(0)], key)] = value
```

Comment:
Is there a reason you did not use a `try`-`catch` here?

Reply:
In retrospect, yes: :upside_down_face:: I believe that more often than not, users will not bother about replacing characters. In these cases, the `if …` clause is faster than a failed `try …`. If there were a default `key_mods` argument, I would change it to a `try …` clause.

Do you want a `try …` here?

Reply:
Not necessarily, I was just wondering given your comment earlier :upside_down_face:

Reply:
I asked for it, didn't I? :stuck_out_tongue_winking_eye: